### PR TITLE
Cast from timestamp to date must honor session timezone

### PR DESCRIPTION
**Pull request by karteekmurthys**
Timestamp to date conversion doesn't honor session timezone when queryConfig `adjust_timestamp_to_session_timezone` to true. This change addresses that.

Testing:
Added unit tests for casting timestamp with timezone to date.

Tested e2e on local machine.
```
presto> select date(from_unixtime(c0)) from (values(0)) t(c0);
   _col0    
------------
 1969-12-31 
```

Resolves: https://github.com/facebookincubator/velox/issues/5596

### CODE REVIEW

File: velox/expression/CastExpr.cpp [hunk 1]
```diff
@@ -147,18 +148,39 @@ VectorPtr castToDate(
...
-          resultFlatVector->set(row, seconds / kSecsPerDay - 1);
-        }
-      });
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      auto sessionTzName = queryConfig.sessionTimezone();
+      if (queryConfig.adjustTimestampToTimezone() && !sessionTzName.empty()) {
+        // locate_zone throws runtime_error if the timezone couldn't be found
+        // (so we're safe to dereference the pointer).
+        auto* timeZone = date::locate_zone(sessionTzName);
+        context.applyToSelectedNoThrow(rows, [&](int row) {
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:157]: This and the loop below differ in only one statement "input.toTimezone(*timeZone)". Maybe abstract a common function for the loop with a boolean template parameter that can be used to control the timezone conversion.

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:157]: Thanks, will address this.

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -119,6 +119,31 @@ VectorPtr castFromDate(
   }
 }
 
+template <bool AdjustForTimeZone>
+void castTimestampToDate(
+    const SelectivityVector& rows,
+    const SimpleVector<Timestamp>* inputVector,
+    exec::EvalCtx& context,
+    FlatVector<int32_t>* resultFlatVector,
+    const date::time_zone* timeZone = NULL) {
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:128]: Nit : Use nullptr.

File: velox/expression/CastExpr.cpp [hunk 3]
```diff
@@ -151,35 +176,12 @@ VectorPtr castToDate(
...
-          if (seconds >= 0 || seconds % kSecsPerDay == 0) {
-            resultFlatVector->set(row, seconds / kSecsPerDay);
-          } else {
-            // For division with negatives, minus 1 to compensate the discarded
-            // fractional part. e.g. -1/86'400 yields 0, yet it should be
-            // considered as -1 day.
-            resultFlatVector->set(row, seconds / kSecsPerDay - 1);
-          }
-        });
+        castTimestampToDate<true>(
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:180]: It might be cleaner to pass const BaseVector& input as a parameter to the callTimestampToDate function and add the line
 auto* inputVector = input.as<SimpleVector<Timestamp>>(); 

as the first line in its function body.

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:180]: Done.

File: velox/expression/CastExpr.cpp (hunk 2 above)
Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp]: Done.

File: velox/expression/CastExpr.cpp [hunk 4]
```diff
@@ -103,6 +103,7 @@ VectorPtr castFromDate(
     }
     case TypeKind::TIMESTAMP: {
       static const int64_t kMillisPerDay{86'400'000};
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:106]: Why is this change needed ? Please remove it if not required.

File: velox/expression/CastExpr.cpp [hunk 5]
```diff
@@ -118,6 +119,32 @@ VectorPtr castFromDate(
   }
 }
 
+template <bool AdjustForTimeZone>
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:122]: Use camelCase adjustForTimezone for non-type template parameters.

File: velox/expression/CastExpr.cpp [hunk 6]
```diff
@@ -145,20 +172,16 @@ VectorPtr castToDate(
       return castResult;
     }
     case TypeKind::TIMESTAMP: {
-      auto* inputVector = input.as<SimpleVector<Timestamp>>();
       static const int32_t kSecsPerDay{86'400};
```
Comment (aditi-pandit) [inline velox/expression/CastExpr.cpp:175]: Nit : This variable isn't needed anymore.

File: velox/expression/tests/CastExprTest.cpp [hunk 7]
```diff
@@ -348,6 +349,26 @@ TEST_F(CastExprTest, timestampToDate) {
       false,
       TIMESTAMP(),
       DATE());
+
+  setTimezone("America/Los_Angeles");
+  testCast<Timestamp, int32_t>(
+      "date",
+      {
+          Timestamp(0, 0),
```
Comment (aditi-pandit) [inline velox/expression/tests/CastExprTest.cpp:357]: The input list is the same in both the tests (which is good) so abstract it as a common variable.

File: velox/expression/CastExpr.cpp (hunk 4 above)
Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:106]: Not related. Removed.

**Review by mbasmanova — COMMENTED** [review]
It would be nice to update documentation for adjust_xxx config to mention how it affects these casts.

https://facebookincubator.github.io/velox/configs.html

File: velox/expression/CastExpr.cpp [hunk 8]
```diff
@@ -118,6 +118,32 @@ VectorPtr castFromDate(
   }
 }
 
+template <bool adjustForTimeZone>
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:121]: The loop body is pretty complex with conditionals and %. One more if is not going to make a big difference. I suggest to remove this template parameter.

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp]: This conditional will be executed for each element of a row. We have seen one condition can make a difference in while executed for each row. Refer: https://github.com/facebookincubator/velox/pull/5455/files#diff-27b090936868d7e88b5053541405438a3b79ba2355bc153789cd48581836e542L50.

I have refactored code to avoid new templated function. Instead using a callback to convert to timezone. Lmk what you think?

Comment (mbasmanova) [reply velox/expression/CastExpr.cpp]: Can you benchmark to double check?

File: velox/expression/CastExpr.cpp [hunk 9]
```diff
@@ -145,10 +145,21 @@ VectorPtr castToDate(
...
+      AdjustForTZFunc adjustTimestampForTimezone = [&](Timestamp& input) {};
+      if (queryConfig.adjustTimestampToTimezone() && !sessionTzName.empty()) {
+        auto* timeZone = date::locate_zone(sessionTzName);
+        adjustTimestampForTimezone = [&](Timestamp& input) {
+          input.toTimezone(*timeZone);
+        };
+      }
       context.applyToSelectedNoThrow(rows, [&](int row) {
         auto input = inputVector->valueAt(row);
+        adjustTimestampForTimezone(input);
```
Comment (mbasmanova) [inline velox/expression/CastExpr.cpp:162]: Callback is usually worse than a check against a const boolean.

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:162]: Agree. I will revert the change.

File: velox/expression/CastExpr.cpp (hunk 8 above)
Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:121]: `if` as constexpr
```
WARNING: Benchmark running in DEBUG mode
============================================================================
[...]benchmarks/basic/SimpleArithmetic.cpp     relative  time/iter   iters/s
============================================================================
castTimestampDate                                          83.34ms     12.00
```

`if` Runtime check
```
WARNING: Benchmark running in DEBUG mode
============================================================================
[...]benchmarks/basic/SimpleArithmetic.cpp     relative  time/iter   iters/s
============================================================================
castTimestampDate                                          85.50ms     11.70
```
Runtime check is 2.16 ms slower than constexpr check.

Comment (mbasmanova) [reply velox/expression/CastExpr.cpp:121]: Thank you for benchmarking. Did you run an existing benchmark or added a new one? If new, would you include it in this PR?

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:121]: I added a new timestamp column under input schema of SimpleArithmeticBenchmark. Included new test as follows:

```
BENCHMARK(castTimestampDate) {
  benchmark->runSmall("cast (e as date)");
  benchmark->runMedium("cast (e as date)");
  benchmark->runLarge("cast (e as date)");
}
```
I will add a new test file for benchmarking cast expr.

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+  RowVectorPtr makeRowVector(vector_size_t size) {
+    VectorFuzzer::Options opts;
+    opts.vectorSize = size;
+    opts.nullRatio = 0;
+    VectorFuzzer fuzzer(opts, pool(), FLAGS_fuzzer_seed);
+
+    std::vector<VectorPtr> children;
+    children.emplace_back(fuzzer.fuzzFlat(TIMESTAMP())); // A
+    opts.nullRatio = 0.5; // 50%
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:56]: This seems very high. I feel it should be 0.2 or 0.3 maybe.

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    fuzzer.setOptions(opts);
+
+    return std::make_shared<RowVector>(
+        pool(), inputType_, nullptr, size, std::move(children));
+  }
+  static constexpr auto kIterationsSmall = 10'000;
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:62]: Nit : Renaming use num in the naming kNumSmallRuns, etc

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+  void runMedium(const std::string& expression) {
+    run(expression, kIterationsMeduim, mediumRowVector_);
+  }
+
+  void runLarge(const std::string& expression) {
+    run(expression, kIterationsLarge, largeRowVector_);
+  }
+
+  // Runs `expression` `times` thousand times.
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:78]: Nit : Is the word thousand required here ?

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+    size_t count = 0;
+    for (auto i = 0; i < times; i++) {
+      count += evaluate(exprSet, input)->size();
+    }
+    return count;
+  }
+
+ private:
+  TypePtr inputType_;
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:93]: Should all these member variables be const ?

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+  explicit SimpleCastBenchmark() : FunctionBenchmarkBase() {}
+
+  RowVectorPtr makeRowVector(vector_size_t size) {
+    VectorFuzzer::Options opts;
+    opts.vectorSize = size;
+    opts.nullRatio = 0;
+    VectorFuzzer fuzzer(opts, pool(), FLAGS_fuzzer_seed);
+
+    std::vector<VectorPtr> children;
+    children.emplace_back(fuzzer.fuzzFlat(TIMESTAMP())); // A
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:45]: Nit : Comment -- what is // A ?

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    opts.nullRatio = 0.3; // 30%
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:46]: Its a bit strange that you set the fuzzer options after using the fuzzer. Shouldn't these be before ?

File: velox/benchmarks/basic/SimpleCastExpr.cpp [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+    fuzzer.setOptions(opts);
+
+    return std::make_shared<RowVector>(
+        pool(), inputType_, nullptr, size, std::move(children));
+  }
+  static constexpr auto kNumSmallRuns = 10'000;
```
Comment (aditi-pandit) [inline velox/benchmarks/basic/SimpleCastExpr.cpp:52]: Nit : Add a newline here.

File: velox/benchmarks/basic/SimpleCastExpr.cpp (hunk 15 above)
Comment (karteekmurthys) [reply velox/benchmarks/basic/SimpleCastExpr.cpp:46]: Sorry, I just moved logic from SimpleArithmetic test to this test. Thanks for pointing out. This has no impact.

File: velox/benchmarks/basic/SimpleCastExpr.cpp (hunk 14 above)
Comment (karteekmurthys) [reply velox/benchmarks/basic/SimpleCastExpr.cpp]: Comment means to say that it is `Col a`. I fixed it.

**Comment by karteekmurthys** [comment]
@mbasmanova would you please review this as well.

**Review by mbasmanova — COMMENTED** [review]
@laithsakka Laith, would you help review and merge this PR?

**Comment by bikramSingh91** [comment]
I will review this either today or at the latest by tomorrow.

**Review by bikramSingh91 — APPROVED** [review]
